packages/cletus/src/helpers/dba-query.ts

```diff
@@ -0,0 +1,1717 @@
+      ctx.dataManagers.set(tableName, manager);
+    }
+    
+    const original = manager.getAll();
+    state = {
+      original,
+      current: [...original],
+      deletedIds: new Set(),
+      updatedIds: new Map(),
+      insertedIds: new Map(),
+    };
+    ctx.tableStates.set(tableName, state);
+  }
+  return state;
+}
+
+/**
+ * Get current records from table state (includes pending changes)
+ */
+function getRecords(state: TableState): DataRecord[] {
+  return state.current;
+}
+
+/**
+ * Add insert operation to table state
+ */
+function addInsert(state: TableState, id: string, fields: Record<string, unknown>): DataRecord {
+  const now = Date.now();
+  const record: DataRecord = {
+    id,
+    created: now,
+    updated: now,
+    fields,
+  };
+  
+  // If this ID was previously deleted, remove the delete
+  if (state.deletedIds.has(id)) {
+    state.deletedIds.delete(id);
+  }
+  
+  state.current.push(record);
+  state.insertedIds.set(id, fields);
+  
+  return record;
+}
+
+/**
+ * Add update operation to table state
+ */
+function addUpdate(state: TableState, id: string, fields: Record<string, unknown>): void {
+  // Find the record in current state
+  const record = state.current.find(r => r.id === id);
+  if (!record) return;
+  
+  // Apply update to current state
+  Object.assign(record.fields, fields);
+  record.updated = Date.now();
+  
+  // If this was a new insert, just update the insert's fields in the map
+  if (state.insertedIds.has(id)) {
+    const existingFields = state.insertedIds.get(id);
+    if (existingFields) {
+      Object.assign(existingFields, fields);
+    }
+  } else {
+    // Merge with existing updates or add new update
+    const existing = state.updatedIds.get(id);
+    if (existing) {
+      Object.assign(existing, fields);
+    } else {
+      state.updatedIds.set(id, { ...fields });
+    }
+  }
+}
+
+/**
+ * Add delete operation to table state
+ */
+function addDelete(state: TableState, id: string): void {
+  // Remove from current state
+  state.current = state.current.filter(r => r.id !== id);
+  
+  // If this was a new insert, just remove the insert
+  if (state.insertedIds.has(id)) {
+    state.insertedIds.delete(id);
+  } else {
+    // Remove any pending updates for this record
+    state.updatedIds.delete(id);
+    
+    // Add delete operation
+    state.deletedIds.add(id);
+  }
+}
+
+/**
+ * Commit all pending changes to disk
+ */
+async function commitChanges(ctx: QueryContext): Promise<{
+  insertedIds: string[];
+  updatedIds: string[];
+  deletedIds: string[];
+}> {
+  const result = {
+    insertedIds: [] as string[],
+    updatedIds: [] as string[],
+    deletedIds: [] as string[],
+  };
+  
+  for (const [tableName, state] of ctx.tableStates) {
+    const manager = ctx.dataManagers.get(tableName);
+    if (!manager) continue;
+    
+    // Process inserts - DataManager.create returns the actual ID
+    for (const [tempId, fields] of state.insertedIds) {
```

Comment:
Do all the create, update, & deletes in a manager.save statement - since these individual calls is costly.